### PR TITLE
feat: replace all streams with Minipass streams

### PR DESCRIPTION
**Pull request by isaacs**
BREAKING CHANGE: This subtly changes the streaming interface of everything in cacache that streams, which is, well, everything in cacache.  Most users will probably not notice, but any code that depended on stream behavior always being deferred until next tick will need to adjust.

The mississippi methods 'to', 'from', 'through', and so on, have been replaced with their Minipass counterparts, and streaming interaction with the file system is done via fs-minipass.

The following modules are of interest here:

- [minipass](http://npm.im/minipass) The core stream library.

- [fs-minipass](http://npm.im/fs-minipass)  Note that the 'WriteStream'   class from fs-minipass is _not_ a Minipass stream, but rather a plain   old EventEmitter that duck types as a Writable.

- [minipass-collect](http://npm.im/minipass-collect) Gather up all the   data from a stream.  Cacache only uses Collect.PassThrough, which is a   basic Minipass passthrough stream which emits a 'collect' event with   the completed data just before the 'end' event.

- [minipass-pipeline](http://npm.im/minipass-pipeline) Connect one or   more streams into a pipe chain.  Errors anywhere in the pipeline are   proxied down the chain and then up to the Pipeline object itself.   Writes go into the head, reads go to the tail.  Used in place of   pump() and pumpify().

- [minipass-flush](http://npm.im/minipass-flush) A Minipass passthrough   stream that defers its 'end' event until after a flush() method has   completed (either calling the supplied callback, or returning a   promise.)  Use in place of flush-write-stream (aka mississippi.to).

Streams from through2, concat-stream, and the behavior provided by end-of-stream are all implemented in Minipass itself.

Features of interest to cacache, which make Minipass a particularly good fit:

- All of the 'endish' events are normalized, so we can just listen on   'end' and know that finish, prefinish, and close will be handled as   well.
- Minipass doesn't waste time [containing   zalgo](https://blog.izs.me/2013/08/designing-apis-for-asynchrony).
- Minipass has built-in support for promises that indicate the end or   error: stream.promise(), stream.collect(), and stream.concat().
- With reliable and consistent timing guarantees, much less   error-checking logic is required.  We can be more confident that an   error is being thrown or emitted in the correct place, rather than in   a callback which is deferred, resulting in a hung promise or   uncaughtException.

The biggest downside of Minipass is that it lacks some of the internal characteristics of node-core streams, which many community modules use to identify streams.  They have no _writableState or _readableState objects, or _read or _write methods.  As a result, the is-stream module (at least, at the time of this commit) doesn't recognize Minipass streams as readable or writable streams.

All in all, the changes required of downstream users should be minimal, but are unlikely to be zero.  Hence the semver major change.

### CODE REVIEW

**Comment by isaacs** [comment]
Also landed #10 on top of it as well, since I was making sure it didn't conflict.

**Comment by isaacs** [comment]
Note: before landing, we should bump the major version of https://github.com/npm/ssri and publish.  Currently, this is depending on the github repo, which is not ideal.

**Comment by isaacs** [comment]
Benchmarks, v13 vs latest:

```
> cacache@12.0.3 benchmarks /home/isaacs/cacache
> node test/benchmarks

================================================
  Comparing to test/cache/last-benchmark.json
================================================
     content.read()
------------------------------------------------
  2565 ops/s @ ~0.390ms/op (-32.76% ±7.51%)
  Sampled 65 in 6.12s.
================================================
     content.read() big data
------------------------------------------------
  60.66 ops/s @ ~16.486ms/op (+47.45% ±3.94%)
  Sampled 64 in 5.85s.
================================================
     content.read.copy() small data
------------------------------------------------
  2734 ops/s @ ~0.366ms/op (-28.51% ±4.39%)
  Sampled 70 in 5.82s.
================================================
     content.read.copy() big data
------------------------------------------------
  145 ops/s @ ~6.908ms/op (-28.58% ±2.89%)
  Sampled 36 in 6.02s.
================================================
     content.read.stream() small data
------------------------------------------------
  1385 ops/s @ ~0.722ms/op (-38.48% ±6.69%)
  Sampled 64 in 5.92s.
================================================
     content.read.stream() big data
------------------------------------------------
  73.50 ops/s @ ~13.606ms/op (+2.69% ±2.80%)
  Sampled 59 in 5.92s.
================================================
     get.byDigest()
------------------------------------------------
  2426 ops/s @ ~0.412ms/op (-19.23% ±4.56%)
  Sampled 68 in 5.81s.
================================================
     get.byDigest() memoized
------------------------------------------------
  28788 ops/s @ ~0.035ms/op (-25.75% ±3.72%)
  Sampled 71 in 5.81s.
================================================
     get.stream.byDigest() small data
------------------------------------------------
  1687 ops/s @ ~0.593ms/op (-32.57% ±3.74%)
  Sampled 71 in 5.91s.
================================================
     get.stream.byDigest() big data
------------------------------------------------
  76.59 ops/s @ ~13.056ms/op (-5.25% ±3.47%)
  Sampled 56 in 5.97s.
================================================
     get.copy.byDigest() small data
------------------------------------------------
  2618 ops/s @ ~0.382ms/op (-10.21% ±2.33%)
  Sampled 70 in 6.03s.
================================================
     get.copy.byDigest() big data
------------------------------------------------
  101 ops/s @ ~9.866ms/op (+14.73% ±34.51%)
  Sampled 49 in 5.97s.
================================================
     index.find cache hit
------------------------------------------------
  3942 ops/s @ ~0.254ms/op (-20.60% ±6.43%)
  Sampled 70 in 5.98s.
================================================
     index.find cache miss
------------------------------------------------
  10031 ops/s @ ~0.100ms/op (-13.71% ±3.13%)
  Sampled 72 in 5.83s.
================================================
     index.insert() different files
------------------------------------------------
  2105 ops/s @ ~0.475ms/op (-6.18% ±4.66%)
  Sampled 68 in 5.87s.
================================================
     index.insert() same file
------------------------------------------------
  2396 ops/s @ ~0.417ms/op (-12.71% ±3.50%)
  Sampled 69 in 5.87s.
================================================
     cacache.put()
------------------------------------------------
  630 ops/s @ ~1.587ms/op (+0.53% ±4.24%)
  Sampled 66 in 5.77s.
================================================
     cacache.put() big data
------------------------------------------------
  14.32 ops/s @ ~69.826ms/op (+31.85% ±3.44%)
  Sampled 37 in 5.63s.
================================================
     cacache.put.stream() 2048 bytes
------------------------------------------------
  446 ops/s @ ~2.242ms/op (-15.01% ±5.76%)
  Sampled 59 in 5.81s.
================================================
     cacache.put.stream() 2048000 bytes
------------------------------------------------
  14.50 ops/s @ ~68.951ms/op (+33.09% ±1.61%)
  Sampled 228 in 35.52s.
================================================
```

Mostly really good!

I intend to dig into why the "put big data", "get big data", and "put.stream 2048000 bytes" benchmarks all got slower.  (Sensing a definite theme there :)

File: lib/content/write.js [hunk 1]
```diff
@@ -44,101 +45,98 @@ function write (cache, data, opts) {
...
-      if (!allDone) {
-        allDone = handleContent(inputStream, cache, opts, errCheck)
+  flush (cb) {
+    this.inputStream.end(() => {
+      if (!this.handleContentP) {
+        const e = new Error('Cache input stream was empty')
+        e.code = 'ENODATA'
+        // empty streams are probably emitting end right away.
+        // defer this one tick by rejecting a promise on it.
+        return Promise.reject(e).catch(cb)
```
Comment (ruyadorno) [inline lib/content/write.js]: 🤔 I noticed there's a missing `stream.emit('error', e)` for this "Cache empty" error here, is that expected to be part of the public API? is it fine if it's just not there anymore? OR is the promise rejection going to trigger that somehow upstream?

File: lib/entry-index.js [hunk 2]
```diff
@@ -158,63 +158,47 @@ module.exports.lsStream = lsStream
...
+                entries.reduce((acc, entry) => {
+                  acc.set(entry.key, entry)
+                  return acc
+                }, new Map())
+              ).then(reduced => {
+                // reduced is a map of key => entry
+                for (const entry of reduced.values()) {
+                  const formatted = formatEntry(cache, entry)
+                  if (formatted) {
+                    stream.write(formatted)
```
Comment (ruyadorno) [inline lib/entry-index.js]: [question] why the change from `stream.push` to `stream.write` in this one?

File: package.json [hunk 3]
```diff
@@ -58,30 +58,34 @@
...
+    "minipass-flush": "^1.0.3",
+    "minipass-pipeline": "^1.1.1",
     "mkdirp": "^0.5.1",
     "move-concurrently": "^1.0.1",
     "p-map": "^3.0.0",
     "promise-inflight": "^1.0.1",
-    "rimraf": "^2.6.3",
-    "ssri": "^6.0.1",
+    "rimraf": "^2.7.1",
+    "ssri": "github:npm/ssri",
```
Comment (ruyadorno) [inline package.json]: 🤔 is this temporary?

File: put.js [hunk 4]
```diff
@@ -44,57 +46,49 @@ function putStream (cache, key, opts) {
...
-  })
-  contentStream.once('error', (err) => {
-    if (erred) {
-      return
+          if (integrity) {
+            pipeline.emit('integrity', integrity)
+          }
+          if (size) {
+            pipeline.emit('size', size)
+          }
```
Comment (ruyadorno) [inline put.js]: this seems different from the original implementation (wasn't checking for `integrity` before `emit` and wasn't emitting `size`) I wonder if that affects anything 🤔

File: test/benchmarks/content.read.js [hunk 5]
```diff
@@ -121,11 +121,11 @@ module.exports = (suite, CACHE) => {
...
-      stream.on('data', () => {})
-      stream.on('error', (err) => deferred.reject(err))
-      stream.on('end', () => {
-        deferred.resolve()
-      })
+      stream.promise().then(
+        () => deferred.resolve(),
+        er => deferred.reject(er)
+      )
+      stream.resume()
```
Comment (ruyadorno) [inline test/benchmarks/content.read.js]: unrelated to these specific lines but reading through these made me realize that we're actually slightly changing the exposed public APIs since the returned streams are now `minipass` streams and might have [differences from Node.js streams](https://www.npmjs.com/package/minipass#differences-from-nodejs-streams) - I wonder if it's worth mentioning something in the `README` file

File: get.js [hunk 6]
```diff
@@ -166,19 +156,24 @@ function getStream (cache, key, opts) {
...
+          size: opts.size == null ? entry.size : opts.size
+        })
       )
+
+      if (opts.memoize) {
+        const memoStream = new Collect.PassThrough()
+        memoStream.on('collect', data => memo.put(cache, entry, data, opts))
+        stream.unshift(memoStream)
+      }
+      stream.unshift(src)
```
[the remark this answers is not in the document]
Comment (isaacs) [reply get.js]: `minipass-pipeline` streams manage a list of streams that are all piped together, and present a single Minipass stream interface on top.  So, writing goes to the `head` of the pipeline, and reading comes from the `tail`.  If the `tail` is not a readable (ie, duplex or PassThrough) stream, then the Pipeline just shoves data into it and that's that.  If the `head` isn't a writable (duplex) then you can't write to the pipeline.

`Pipeline.push(stream)` adds it to the end of the pipeline.  `Pipeline.unshift(stream)` ads it to the start.

So, these three are all equivalent:

```js
// provide to constructor
p = new Pipeline(reader, transform, writer)

// using push
p = new Pipeline()
p.push(reader) // {reader}
p.push(transform) // {reader, transform}
p.push(writer) // {reader, transform, writer}

// using unshift
p = new Pipeline()
p.unshift(writer) // {writer}
p.unshift(transform) // {transform, writer}
p.unshift(reader) // {reader, transform, writer}

// using all three things!
p = new Pipeline(transform) // {transform}
p.unshift(reader) // {reader, transform}
p.push(writer) // {reader, transform, writer}
```

In other words, a Pipeline is pushing _streams_ into the set, using push/unshift like an Array would, to add to the ordered list.

Node-core Readable streams have a "push" method which is a very different thing.  That method is for pushing _data_ into the stream itself.  And, like the push/unshift paradigm from arrays, `s.push(data)` adds the data into the end of the queue, but `s.unshift(data)` adds the data to the _start_ of the queue, which leads to some very confusing and weird scenarios that are really complicated to manage.

File: lib/content/write.js (hunk 1 above)
Comment (isaacs) [reply lib/content/write.js]: `minipass-flush` streams will emit an error if the flush cb is called with an error.  Since Minipass streams can emit `end` immediately when `end()` is called, and that would mean that the caller doesn't get a chance to catch the error, it's deferred by the `Promise.reject(e).catch(cb)` rather than doing `return cb(er)` synchronously.  I guess I could have also done `Promise.resolve(e).then(cb)`, but both are slightly faster than `setTimeout(() => cb(er))` or `setImmediate(() => cb(er))`.

File: lib/entry-index.js (hunk 2 above)
Comment (isaacs) [reply lib/entry-index.js]: See above about Node core's 'Readable.push()` method.  Base Minipass streams don't have a "push()".  But the base Minipass class is a PassThrough stream, so `stream.write()` queues the data up to be emitted by a `'data'` event (with appropriate buffering, etc.)

File: package.json (hunk 3 above)
Comment (isaacs) [reply package.json]: Yes :)

Before this can land (or at least, before it can be published) we'll need to merge in ssri's v7 branch, and publish the new semver major.

Comment (isaacs) [reply package.json]: See: https://github.com/npm/ssri/pull/3

File: put.js (hunk 4 above)
Comment (isaacs) [reply put.js]: Good catch.  I don't think it would cause a problem, and we almost always emit size along with integrity.  Probably fine?  At least, consistency is good.

File: test/benchmarks/content.read.js (hunk 5 above)
Comment (isaacs) [reply test/benchmarks/content.read.js]: Yeah, doc update would be good.  As it was, Cacache returned a mix of various mississippi streams, but most of those inherited from stream.Readable or stream.Duplex, so they would function more or less identically to node-core streams.

File: get.js [hunk 7]
```diff
@@ -120,43 +123,30 @@ function getDataSync (byDigest, cache, key, opts) {
...
+  const stream = new Minipass()
+  stream.on('newListener', function (ev, cb) {
+    ev === 'metadata' && cb(memoized.entry.metadata)
+    ev === 'integrity' && cb(memoized.entry.integrity)
+    ev === 'size' && cb(memoized.entry.size)
+  })
+  stream.end(memoized.data)
+  return stream
+}
+
```
[the remark this answers is not in the document]
Comment (isaacs) [reply get.js]: That'd be good, yes.

File: get.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (isaacs) [reply get.js]: Ok, well... I wrote a whole explainer here, and then GitHub dropped it, so you get the short version:

`stream` here is a [minipass-pipeline](http://npm.im/minipass-pipeline).  `push` adds a stream to the end of the pipeline.  `unshift` adds a stream to the start of the pipeline.

This is different from node-core's `stream.push(chunk)` which queues a chunk of _data_ top be emitted.  In Minipass lingo, you'd use `this.write(data)`, since they're passthrough streams (or `super.write(data)` if the `write` method has been overridden, which is pretty common.)

File: lib/content/write.js (hunk 1 above)
Comment (isaacs) [reply lib/content/write.js]: This line will cause hte `cb` to be called with an error as its first argument.  [minipass-flush](http://npm.im/minipass-flush) streams will emit an error if the `flush(cb)` callback is called with an error.

File: lib/entry-index.js (hunk 2 above)
Comment (isaacs) [reply lib/entry-index.js]: See above about `Minipass.write()`.  Same logical meaning, different stream implementation :)

File: put.js (hunk 4 above)
Comment (isaacs) [reply put.js]: Huh, that is weird.  We emit `size` almost everywhere else.  I like the consistency, it's probably fine.

File: test/benchmarks/content.read.js (hunk 5 above)
Comment (isaacs) [reply test/benchmarks/content.read.js]: Agreed.

File: package.json (hunk 3 above)
Comment (isaacs) [reply package.json]: Yes, hopefully it won't stay that way for long :)

https://github.com/npm/ssri/pull/3

File: get.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (isaacs) [reply get.js]: ok, now the original thing I wrote is back.

Internet gremlins.

**Comment by isaacs** [comment]
Updated benchmarks with the `fs-minipass` usage and a few other slight tweaks when reading large data.

```
================================================
  Comparing to test/cache/last-benchmark.json
================================================
     content.read()
------------------------------------------------
  1841 ops/s @ ~0.543ms/op (+3.25% ±7.15%)
  Sampled 63 in 6.05s.
================================================
     content.read() big data
------------------------------------------------
  47.13 ops/s @ ~21.219ms/op (-14.36% ±6.65%)
  Sampled 56 in 6.02s.
================================================
     content.read.copy() small data
------------------------------------------------
  2198 ops/s @ ~0.455ms/op (-41.29% ±3.85%)
  Sampled 69 in 5.93s.
================================================
     content.read.copy() big data
------------------------------------------------
  107 ops/s @ ~9.324ms/op (+17.92% ±3.57%)
  Sampled 45 in 5.78s.
================================================
     content.read.stream() small data
------------------------------------------------
  1379 ops/s @ ~0.725ms/op (-68.14% ±5.26%)
  Sampled 62 in 6.01s.
================================================
     content.read.stream() big data
------------------------------------------------
  81.22 ops/s @ ~12.312ms/op (-65.07% ±4.58%)
  Sampled 60 in 5.92s.
================================================
     get.byDigest()
------------------------------------------------
  2000 ops/s @ ~0.500ms/op (-49.19% ±5.02%)
  Sampled 69 in 6.08s.
================================================
     get.byDigest() memoized
------------------------------------------------
  24761 ops/s @ ~0.040ms/op (-52.52% ±5.31%)
  Sampled 66 in 6.02s.
================================================
     get.stream.byDigest() small data
------------------------------------------------
  1650 ops/s @ ~0.606ms/op (-68.41% ±5.32%)
  Sampled 65 in 5.93s.
================================================
     get.stream.byDigest() big data
------------------------------------------------
  86.52 ops/s @ ~11.558ms/op (-65.92% ±4.67%)
  Sampled 62 in 6.10s.
================================================
     get.copy.byDigest() small data
------------------------------------------------
  2341 ops/s @ ~0.427ms/op (-35.33% ±3.40%)
  Sampled 62 in 5.79s.
================================================
     get.copy.byDigest() big data
------------------------------------------------
  104 ops/s @ ~9.642ms/op (+15.13% ±2.76%)
  Sampled 47 in 5.78s.
================================================
     index.find cache hit
------------------------------------------------
  2656 ops/s @ ~0.377ms/op (+21.82% ±5.49%)
  Sampled 59 in 5.76s.
================================================
     index.find cache miss
------------------------------------------------
  8133 ops/s @ ~0.123ms/op (+11.70% ±3.54%)
  Sampled 66 in 6.06s.
================================================
     index.insert() different files
------------------------------------------------
  1521 ops/s @ ~0.658ms/op (+25.79% ±4.32%)
  Sampled 62 in 5.74s.
================================================
     index.insert() same file
------------------------------------------------
  1873 ops/s @ ~0.534ms/op (+11.90% ±4.22%)
  Sampled 61 in 5.88s.
================================================
     cacache.put()
------------------------------------------------
  569 ops/s @ ~1.758ms/op (+6.39% ±4.37%)
  Sampled 64 in 5.81s.
================================================
     cacache.put() big data
------------------------------------------------
  12.21 ops/s @ ~81.899ms/op (+14.59% ±6.16%)
  Sampled 57 in 5.91s.
================================================
     cacache.put.stream() 2048 bytes
------------------------------------------------
  435 ops/s @ ~2.298ms/op (-6.97% ±5.95%)
  Sampled 61 in 5.90s.
================================================
     cacache.put.stream() 2048000 bytes
------------------------------------------------
  11.33 ops/s @ ~88.281ms/op (+18.73% ±2.44%)
  Sampled 310 in 35.85s.
================================================
```

At this point, the slower (positive) numbers are within the range of what I get from repeated runs on a single branch, so I think it's more owing to the disk on my linux box than any code changes.  The significant speed increases (negative 40% or more) are pretty consistent, though.  So I'd say it's a win.

File: get.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ruyadorno) [reply get.js]: oooh I get it, just mixed up the minipass api with node's Readable/Writable 😅 

I'm really lucky I got the long version, very instructive, thank you so much for it ❤️

File: lib/content/write.js (hunk 1 above)
Comment (ruyadorno) [reply lib/content/write.js]: I see! could be nice to have a note or example of that in `minipass-flush` README - btw I can't find it's source if you I can point out I can send a PR myself 😊

File: put.js (hunk 4 above)
Comment (ruyadorno) [reply put.js]: sounds good 👍
